File: python/src/core/segment.py
from core.OpenFigure import OpenFigure
from core.point import Point
from core.point2d import Point2D

class Segment(OpenFigure):
	def __init__(self, point1, point2):
		self.start = point1
		self.finish = point2

	def __str__(self):
		return "Segment: start = {0}; finish = {1}.".format(self.start, self.finish)

	def getStart(self):
		return self.start

	def getFinish(self):
		return self.finish

	def setStart(self, newStart):
		result = Segment(self.start, self.finish)
		result.start = newStart
		return result

	def setFinish(self, newFinish):
		result = Segment(self.start, self.finish)
		result.start = newFinish
		return result

	def length(self):
		return Point.abs(Point2D.sub(self.finish, self.start))

	def shift(self, point):
		result = Segment(self.start, self.finish)
		result.start = result.start.add(point)
		result.finish = result.finish.add(point)
		return result

	def rot(self, phi):
		result = Segment(self.start, self.finish)
		result.start = result.start.rot(phi)
		result.finish = result.finish.rot(phi)
		return result

	def symAxis(self, i):
		if i < 0:
			raise Exception("Ошибка: отрицательное пространство")
		if self.start.dim < i:
			raise Exception("Ошибка: имеет пространство меньшее, чем нужное")
		result = Segment(self.start, self.finish)
		result.start = result.start.symAxis(i)
		result.finish = result.finish.symAxis(i)
		return result

	def intersectSeg(seg1, seg2):
		a = seg1.start
		b = seg1.finish
		c = seg2.start
		d = seg2.finish
		def ccw(a, b, c):
			return (c.getX(1) - a.getX(1)) * (b.getX(0) - a.getX(0)) > (b.getX(1) - a.getX(1)) * (c.getX(0) - a.getX(0))
		return ccw(a, c, d) != ccw(b, c, d) and ccw(a, b, c) != ccw(a, b, d)

	def intersectCir(seg, cir):
		return Point2D.sub(seg.start, cir.getP()).abs() < cir.getR() or Point2D.sub(seg.finish, cir.getP()).abs() < cir.getR()

	def cross(self, figure):
		if figure.__class__.__name__ == "Segment":
			return Segment.intersectSeg(self, figure)
		elif figure.__class__.__name__ == "Circle":
			return Segment.intersectCir(self, figure)
		elif figure.__class__.__name__ == "Polyline":
			pprev = figure.p[0]
			for point in figure.p[1:]:
				seg = Segment(pprev, point)
				if self.cross(seg):
					return True
				pprev = point
			return False

File: csharp/WorkOOP/Segment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Versioning;
using System.Text;
using System.Threading.Tasks;

namespace WorkOOP
{
    // Реализация класса Segment – отрезок.
    internal class Segment: OpenFigure
    {
        // start – начальная точка отрезка.
        protected Point2D start;
        // finish – конечная точка отрезка.
        protected Point2D finish;

        /*
         * Segment – конструктор, описывающий инициализацию объекта. На вход принимает две точки – начало и конец.
         */
        public Segment(Point2D s, Point2D f)
        {
            start = s;
            finish = f;
        }

        /*
         * Реализация всех функций, кроме square интерфейса OpenFigure
         */
        public Point2D getStart() => start;
        public Point2D setStart(Point2D a) => start = a;
        public Point2D getFinish() => finish;
        public void setFinish(Point2D a) => finish = a;
        public override double length() => Point2D.sub(finish, start).abs();
        public override Segment shift(Point2D a) { start.add(a); finish.add(a); return this; }
        public override Segment rot(double phi) { start.rot(phi); finish.rot(phi); return this; }
        public override Segment symAxis(int i) { start.symAxis(i); finish.symAxis(i); return this; }
        
        public override bool cross(IShape i) { 
            if (i.GetType() == typeof(Segment)) {
                return counterclockwise(this.start, (i as Segment).start, (i as Segment).finish)
                    != counterclockwise(this.finish, (i as Segment).start, (i as Segment).finish) &&
                    counterclockwise(this.start, this.finish, (i as Segment).start) !=
                    counterclockwise(this.start, this.finish, (i as Segment).finish);
            }
            else if (i.GetType() == typeof(Circle)) {
                return (Point2D.sub(start, (i as Circle).getP())).abs() < (i as Circle).getR() 
                    || (Point2D.sub(finish, (i as Circle).getP())).abs() < (i as Circle).getR();
            }
            else
            {
                Point2D[] pts = null;
                if (i.GetType() == typeof(Polyline)) pts = (i as Polyline).getP();
                else if (i.GetType() == typeof(NGon)) pts = (i as NGon).getP();
                else if (i.GetType() == typeof(TGon)) pts = (i as TGon).getP();
                else if (i.GetType() == typeof(QGon)) pts = (i as QGon).getP();
                else if (i.GetType() == typeof(Rectangle)) pts = (i as Rectangle).getP();
                else if (i.GetType() == typeof(Trapeze)) pts = (i as Trapeze).getP();
                if (pts == null) throw new Exception("No points have been passed down");
                Point2D prev = pts[0];
                foreach (Point2D pt in pts)
                {
                    if (new Segment(prev, pt).cross(this)) return true;
                    prev = pt;
                }                
                if ((i.GetType().IsSubclassOf(typeof(NGon)) || i.GetType() == typeof(NGon)) && new Segment(pts[0], pts[pts.Length - 1]).cross(this)) return true;
                return false;
            }
            return false;
        }

        public bool counterclockwise(Point2D a, Point2D b, Point2D c) => (c.getX()[1] - a.getX()[1]) * (b.getX()[0] - a.getX()[0])
                                                                         > (b.getX()[1] - a.getX()[1]) * (c.getX()[0] - a.getX()[0]);
        public override String toString() => "start: " + start.toString() + ", finish:" + finish.toString();
    }
}
